lib/spack/spack/package.py

```diff
@@ -1558,25 +1563,59 @@ def test_process():
                     shutil.rmtree(testdir)
                 fsys.mkdirp(testdir)
 
-                # copy test data into testdir/data
-                datadir = Prefix(self.package_dir).test
-                if os.path.isdir(datadir):
-                    shutil.copytree(datadir, testdir.data)
+                # cd to test directory
+                os.chdir(testdir)
+
+                # run test methods from the package and all virtuals it provides
+                # virtuals have to be deduped by name
+                v_names = list(set([vspec.name
+                                    for vspec in self.virtuals_provided]))
+
+                # hack for compilers that are not dependencies (yet)
+                # TODO: this all eventually goes away
+                if self.name in ('gcc', 'intel', 'intel-parallel-studio', 'pgi'):
+                    v_names.extend(['c', 'cxx', 'fortran'])
+                if self.spec.satisfies('llvm+clang'):
+                    v_names.extend(['c', 'cxx'])
+
+                test_specs = [self.spec] + [spack.spec.Spec(v_name)
+                                            for v_name in v_names]
+                for spec in test_specs:
+                    # Fail gracefully if a virtual has no package/tests
+                    try:
+                        spec_pkg = spec.package
+                    except spack.repo.UnknownPackageError:
+                        continue
 
-                try:
-                    os.chdir(testdir)
-                    self.test()
-                    if self.test_failures:
-                        raise TestFailure(self.test_failures)
-
-                    # cleanup test directory on success
-                    if remove_directory:
-                        shutil.rmtree(testdir)
-                        if not os.listdir(self.test_stage):
-                            shutil.rmtree(self.test_stage)
-                finally:
-                    # reset debug level
-                    tty.set_debug(old_debug)
+                    # copy test data into testdir/data
+                    data_source = Prefix(spec_pkg.package_dir).test
+                    data_dir = os.path.join(testdir.data, spec.name)
+                    if os.path.isdir(data_source):
+                        shutil.copytree(data_source, data_dir)
```

Comment:
are tests always owned by the person running them?  Any permissions here?

Reply:
There shouldn't be any permissions issue here. The tests are being copied from the spack package dir (which shouldn't have permissions issues), and the copies will be owned by the person running the tests

Reply:
👍